From 382663373d63b7e1ee1e3e16e844fe7a4ae0a3c3 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:05:09 -0400
Subject: [fs] btrfs: test for valid bdev before kobj removal in btrfs_rm_device

Message-id: <1407179109-13714-423-git-send-email-sandeen@redhat.com>
Patchwork-id: 87293
O-Subject: [RHEL7.1 PATCH 422/422] btrfs: test for valid bdev before kobj removal in btrfs_rm_device
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

commit 0bfaa9c5cb479cebc24979b384374fe47500b4c9
Author: Eric Sandeen <sandeen@redhat.com>
Date:   Mon Jul 7 12:34:49 2014 -0500

    btrfs: test for valid bdev before kobj removal in btrfs_rm_device

    commit 99994cd btrfs: dev delete should remove sysfs entry
    added a btrfs_kobj_rm_device, which dereferences device->bdev...
    right after we check whether device->bdev might be NULL.

    I don't honestly know if it's possible to have a NULL device->bdev
    here, but assuming that it is (given the test), we need to move
    the kobject removal to be under that test.

    (Coverity spotted this)

    Signed-off-by: Eric Sandeen <sandeen@redhat.com>
    Signed-off-by: Chris Mason <clm@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/volumes.c b/fs/btrfs/volumes.c
index 7995fb9..43b12cd 100644
--- a/fs/btrfs/volumes.c
+++ b/fs/btrfs/volumes.c
@@ -1680,11 +1680,11 @@ int btrfs_rm_device(struct btrfs_root *root, char *device_path)
  if (device->bdev == root->fs_info->fs_devices->latest_bdev)
   root->fs_info->fs_devices->latest_bdev = next_device->bdev;
 
- if (device->bdev)
+ if (device->bdev) {
   device->fs_devices->open_devices--;
-
- /* remove sysfs entry */
- btrfs_kobj_rm_device(root->fs_info, device);
+  /* remove sysfs entry */
+  btrfs_kobj_rm_device(root->fs_info, device);
+ }
 
  call_rcu(&device->rcu, free_device);
 
-- 
1.7.1